src/Azure.Functions.Cli/Common/AuthManager.cs

```diff
@@ -0,0 +1,312 @@
+            MiddlewareAuthSettings.SetAuthSetting("WEBSITE_AUTH_SIGNING_KEY", signingKey);
+            MiddlewareAuthSettings.SetAuthSetting("WEBSITE_AUTH_ALLOWED_AUDIENCES", serializedReplyUrls);
+            MiddlewareAuthSettings.Commit();
+        }
+
+        public Dictionary<string, string> CreateAuthSettingsToPublish(string homepage, string clientId, string clientSecret, string tenant, List<string> replyUrls)
+        {
+            // The 'allowedAudiences' setting of /config/authsettings is of the form ["{replyURL1}", "{replyURL2}"]
+            string serializedArray = JsonConvert.SerializeObject(replyUrls, Formatting.Indented);
+
+            var authSettingsToPublish = new Dictionary<string, string>();
+            authSettingsToPublish.Add("allowedAudiences", serializedArray);
+            authSettingsToPublish.Add("isAadAutoProvisioned", "True");
+            authSettingsToPublish.Add("clientId", clientId);
+            authSettingsToPublish.Add("clientSecret", clientSecret);
+            authSettingsToPublish.Add("defaultProvider", "0"); // 0 corresponds to AzureActiveDirectory
+            authSettingsToPublish.Add("enabled", "True");
+            authSettingsToPublish.Add("issuer", "https://sts.windows.net/" + tenant + "/");
+            authSettingsToPublish.Add("runtimeVersion", "1.0.0");
+            authSettingsToPublish.Add("tokenStoreEnabled", "True");
+            authSettingsToPublish.Add("unauthenticatedClientAction", "1"); // Corresponds to AllowAnonymous
+
+            return authSettingsToPublish;
+        }
+    
+        private static async Task<bool> PublishAuthSettingAsync(Site functionApp, string accessToken, Dictionary<string, string> authSettings)
+        {
+            functionApp.AzureAuthSettings = authSettings;
+            var result = await AzureHelper.UpdateFunctionAppAuthSettings(functionApp, accessToken);
+            if (!result.IsSuccessful)
+            {
+                ColoredConsole
+                    .Error
+                    .WriteLine(ErrorColor("Error updating app settings:"))
+                    .WriteLine(ErrorColor(result.ErrorResult));
+                return false;
+            }
+            return true;
+        }
+
+        static string ComputeSha256Hash(string rawData)
+        {
+            // Create a SHA256   
+            using (SHA256 sha256Hash = SHA256.Create())
+            {
+                // ComputeHash - returns byte array  
+                byte[] bytes = sha256Hash.ComputeHash(Encoding.UTF8.GetBytes(rawData));
+                // Convert byte array to a string   
+                StringBuilder builder = new StringBuilder();
+                for (int i = 0; i < bytes.Length; i++)
+                {
+                    builder.Append(bytes[i].ToString("x2"));
+                }
+                return builder.ToString();
+            }
+        }
+
+        public static string GeneratePassword(int length)
```

Comment:
can you explain a bit what this password is for and what its format is like?

Reply:
This generates an AAD application client secret. An example of what it generates is: "77S2ca0Gwf8mTr7s85ZEqe7dcxKnF4DdyMAF5gNFaTBY6gBEDtcPwQWlTP3KWNElvld9eFyXZdvFfmEnqmrHQA7ki0$LRQ8TL3uLrj6S$F0PRtRYnZ8Dp9xReKNQK9z"